crates/merkle-tree/src/lib.rs

```diff
@@ -1,4 +1,31 @@
+use stark_hash::Felt;
+
 pub mod contract_state;
 pub mod merkle_node;
 pub mod merkle_tree;
 pub mod state_tree;
+
+/// Hashing function used by a particular merkle tree implementation.
+pub trait Hash {
+    fn hash(left: Felt, right: Felt) -> Felt;
+}
+
+/// Implements [Hash] for the [StarkNet Pedersen hash](stark_hash::stark_hash).
+#[derive(Debug, Clone, Copy)]
+pub struct PedersenHash {}
+
+impl Hash for PedersenHash {
+    fn hash(left: Felt, right: Felt) -> Felt {
+        stark_hash::stark_hash(left, right)
+    }
+}
+
+/// Implements [Hash] for the StarkNet Poseidon hash.
+///
+/// TODO: add once hash is implemented.
+struct PoseidonHash;
+impl crate::Hash for PoseidonHash {
+    fn hash(_left: stark_hash::Felt, _right: stark_hash::Felt) -> stark_hash::Felt {
+        unimplemented!("Hash function still needs to be implemented");
+    }
+}
```

Comment:
Does anyone know of a better way to do this? Ideally, instead of a new trait, the merkle tree would be generic over `Fn(Felt, Felt) -> Felt` but the ability to call these in code is nightly.. But maybe I'm missing something?

e.g. this is not possible on stable:
```rust
fn example<Hash: Fn(Felt, Felt) -> Felt>(a: Felt, b: Felt) -> Felt {
    Hash::call(a, b)
}
```
Is there some other way to invoke it maybe? Closures somehow?

Reply:
Nevermind, you need fn_traits, which is indeed not yet stable.